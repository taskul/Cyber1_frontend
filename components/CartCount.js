import { CSSTransition, TransitionGroup } from 'react-transition-group';
import styled from 'styled-components';

const Dot = styled.div`
  padding: 0.5rem;
  line-height: 2rem;
  min-width: 2rem;
  margin-left: 3rem;
  background: var(--yellow, #e5d283);
  color: var(--blackBg, #000);
  /* When numbers change they move within the component and using this we set all numbers to have the same width*/
  font-feature-settings: 'tnum';
  font-variant-numeric: tabular-nums;
`;

const AnimationStyles = styled.span`
  /* we are using these classes to create custom style and animation for the CSSTransitions we are using below */
  position: relative;
  .count {
    display: block;
    position: relative;
    transition: transform 0.4s;
    /* when we are rotating an element we want the back face of it to be hidden so we don't see it in animation */
    backface-visibility: hidden;
  }
  /* The reason we have two styling specificaitons enter and exit is because elements switch during re-render and so we style the component that is entering and the one exiting */
  .count-enter {
    /* rotate 0.5turn is half a turn */
    box-shadow: 0 0 35px #a0a14a;
    transform: scale(4) rotateX(0.5turn);
  }
  .count-enter-active {
    transform: rotateX(0);
  }
  .count-exit {
    /* we put it on top of the new element being rendered. */
    top: 0;
    position: absolute;
    transform: rotateX(0);
    backface-visibility: hidden;
  }
  .count-exit-active {
    transform: scale(4) rotateX(0.5turn);
  }
`;

export default function CartCount({ count }) {
  return (
    <AnimationStyles>
      <TransitionGroup>
        {/* 
      using CSSTransition for some animations
      also 400 milliseconds for time 
      */}
        <CSSTransition
          unmountOnExit
          classNames="count"
          key={count}
          timeout={{ enter: 400, exit: 400 }}
        >
          <Dot>{count}</Dot>
        </CSSTransition>
      </TransitionGroup>
    </AnimationStyles>
  );
}
