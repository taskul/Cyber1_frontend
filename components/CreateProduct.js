import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import Router from 'next/router';
import useForm from '../lib/useForm';
import Form from './styles/Form';
import DisplayError from './ErrorMessage';
import { ALL_PRODUCTS_QUERY } from './Products';

// to be able to dynamically pass data to our mutation we need to give it a name in order for us to be able to reference it
const CREATE_PRODUCT_MUTATION = gql`
  mutation CREATE_PRODUCT_MUTATION(
    # All of this is GraphQL notation
    # which varilabe are getting passed in? and what types are they
    # ! and the end means it is required
    $name: String!
    $description: String!
    $price: Int!
    $image: Upload!
  ) {
    createProduct(
      data: {
        name: $name
        description: $description
        price: $price
        status: "AVAILABLE"
        # How can we create a relationship if we don't have a photo ID?
        # In keystone specifically we can do nested creates and link them as a relationship all from one query.
        # what we are saying here is create a relationship and behind the scenes also create a new item with this image and altText
        photo: { create: { image: $image, altText: $name } }
      }
    ) {
      id
      price
      description
      name
    }
  }
`;

export default function CreateProduct() {
  // we'll use our custom hook to manage form state
  const { inputs, handleChange, resetForm, clearForm } = useForm({
    image: '',
    name: '',
    price: 0,
    description: '',
  });

  // we''ll use useMutation hook
  // first argument is the actual mutation
  // seond is any data we want to pass along, in our case it is all of the values from input fields
  // The first item useMutation returns is the actual function that will run the mutation
  // the second is an object that is the same as what our query returns
  const [createProduct, { data, error, loading }] = useMutation(
    CREATE_PRODUCT_MUTATION,
    {
      // because the names in the inputs match the names of the mutation query the inputs object will be descructured
      variables: inputs,
      // we import the ALL_PRODUCTS_QUERY in order to refetch all products, because when we add a new product the products page is not automatically updated since the Products page is cached and the browser just pulls up the cached version, so what we do is we refetch all of the procuts which updates the product page and will include a new produst. Otherwise user has to manually refresh the page to get all products
      // we can also pass variable to the query if we need to, but we don't have any
      //   refetchQueries: [{ query: ALL_PRODUCTS_QUERY, variables}],
      refetchQueries: [{ query: ALL_PRODUCTS_QUERY }],
    }
  );
  // createProduct when it runs it will go off to the backend and run the mutation for us.
  return (
    <Form
      onSubmit={async (e) => {
        e.preventDefault();
        // submit the input fields to the backend
        const res = await createProduct();
        clearForm();
        // redirect to next page
        Router.push({
          pathname: `/product/${res.data.createProduct.id}`,
        });
      }}
    >
      {/* manages displaying of errors */}
      <DisplayError error={error} />
      {/* we can wrap everyting in fieldset in order to be able to disable all fields input. This way while we are sending data to server the user can't change data in input fields 
        fieldset disabled
        we have it as disabled={loading}, meaning if the form is loading, disable changing input fields
    
        however we can also use more accessible attribute aria-busy
        fieldset aria-busy which will play an animation of the loading bar above the input fields to indicate that something is happening
        aria-busy={loading} means the aria-busy animation will activte when the data is loading to backend
      */}
      <fieldset disabled={loading} aria-busy={loading}>
        <label htmlFor="image">
          Image
          {/* image doesn't need value or placeholder */}
          <input
            required
            type="file"
            id="image"
            name="image"
            onChange={handleChange}
          />
        </label>
        <label htmlFor="name">
          Name
          <input
            type="text"
            id="name"
            name="name"
            placeholder="Name"
            value={inputs.name}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="price">
          Price
          <input
            type="number"
            id="price"
            name="price"
            placeholder="0"
            value={inputs.price}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="description">
          Description
          <textarea
            id="description"
            name="description"
            placeholder="Description"
            value={inputs.description}
            onChange={handleChange}
          />
        </label>
        <button type="submit">Add Product</button>
      </fieldset>
    </Form>
  );
}

export { CREATE_PRODUCT_MUTATION };
