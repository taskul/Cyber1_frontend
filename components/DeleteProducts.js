import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';

// our mutation named DELETE_PRODUCT_MUTATION takes an id of a type ID that is required !
// it will return id and name after deletion
const DELETE_PRODUCT_MUTATION = gql`
  mutation DELETE_PRODUCT_MUTATION($id: ID!) {
    deleteProduct(id: $id) {
      id
      name
    }
  }
`;

// Apollo has a really nice API called Evict, and it basically just takes the deleted item out of the cache, react then will notice the change and re-render taking that item out.
// the way that this works is we create an update function that we pass along with our useMutation
// we have two arguments cache which gives us access to Apllo cache, and payload which is what useMutation returns, in our case it is "id" and "name"
function update(cache, payload) {
  // here we don't have reference to the item so the way we find it is by using cache.identify. since data.deleteProduct returns an ID we then can find that item using the ID passed in
  // if we console.log(payload) we'll find data object in there
  cache.evict(cache.identify(payload.data.deleteProduct));
}

export default function DeleteProduct({ id, children }) {
  const [deleteProduct, { loading, error }] = useMutation(
    DELETE_PRODUCT_MUTATION,
    {
      variables: { id },
      update,
    }
  );
  return (
    <button
      type="button"
      disabled={loading}
      style={{ cursor: 'pointer' }}
      onClick={() => {
        // we'll use browser native confirmation, if user confirms then delete the item
        if (confirm('Are you sure you want to delete this item?')) {
          deleteProduct().catch((err) => alert(err.message));
        }
      }}
    >
      {/* the reason we put children in there is because it is a flexible button
      that will accept any text we place in it. */}
      {children}
    </button>
  );
}
