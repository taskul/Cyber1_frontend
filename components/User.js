import { gql, useQuery } from '@apollo/client';

// the reason we are using authenticated item and not user is because Keystone makes it sort of agnostic as to waht this thing is called. When we set up our schema, we told it to use user as the authenticated item, but it could be authenticated company, etc.
// we can not query User field on type AuthenticatedItem
// it returns a union, and union means it could return multiple types of values, and in out case it could return user.
export const CURRENT_USER_QUERY = gql`
  query {
    authenticatedItem {
      ... on User {
        id
        email
        name
        cart {
          id
          quantity
          product {
            id
            price
            name
            description
            photo {
              image {
                publicUrlTransformed
              }
            }
          }
        }
      }
    }
  }
`;
export function useUser() {
  const { data } = useQuery(CURRENT_USER_QUERY);
  return data?.authenticatedItem;
}
