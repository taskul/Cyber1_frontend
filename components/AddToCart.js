import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import { CURRENT_USER_QUERY } from './User';

const ADD_TO_CART_MUTATION = gql`
  mutation ADD_TO_CART_MUTATION($id: ID!) {
    addToCart(productId: $id) {
      id
    }
  }
`;

export default function AddToCart({ id }) {
  const [addToCart, { loading }] = useMutation(ADD_TO_CART_MUTATION, {
    variables: { id },
    // refetching current user pulls up all of the info about the user including the items they have added to their cart, this is how we update the number of items in the cart.
    refetchQueries: [{ query: CURRENT_USER_QUERY }],
  });
  return (
    <button
      type="button"
      disabled={loading}
      style={{ cursor: 'pointer' }}
      onClick={addToCart}
    >
      Add{loading && 'ing'} To Cart &#128092;
    </button>
  );
}
