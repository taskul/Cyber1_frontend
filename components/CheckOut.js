import {
  CardElement,
  Elements,
  useElements,
  useStripe,
} from '@stripe/react-stripe-js';
import { loadStripe } from '@stripe/stripe-js';
import styled from 'styled-components';
import { useState } from 'react';
import nProgress from 'nprogress';
import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import { useRouter } from 'next/dist/client/router';
import SickButton from './styles/SickButton';
import { useCart } from '../lib/cartState';
import { CURRENT_USER_QUERY } from './User';

const CheckoutFormStyles = styled.form`
  box-shadow: 0 1px 2px 2px rgba(0, 0, 0, 0.04);
  border: 1px solid rgba(0, 0, 0, 0.06);
  border-radius: 5px;
  padding: 1rem;
  display: grid;
`;

const CREATE_ORDER_MUTATION = gql`
  mutation CREATE_ORDER_MUTATION($token: String!) {
    checkout(token: $token) {
      id
      charge
      total
      items {
        id
        name
      }
    }
  }
`;

const stripeLib = loadStripe(`${process.env.NEXT_PUBLIC_STRIPE_KEY}`);

function CheckOutForm() {
  const [error, setError] = useState();
  const [loading, setLoading] = useState(false);
  // when we declare this it needs to have access to Elements provider, which is why we are wrapping whole CheckOutForm in the Elements component below
  const stripe = useStripe();
  const elements = useElements();
  const router = useRouter();
  const { closeCart } = useCart();
  const [checkout, { error: graphqlError }] = useMutation(
    CREATE_ORDER_MUTATION,
    {
      // after user buys something, here we refetch the user which would also pull user cart information which would be empty at this point.
      refetchQueries: [{ query: CURRENT_USER_QUERY }],
    }
  );

  async function handleSubmit(e) {
    // 1 stop form from submiting and turn the loader on
    e.preventDefault();
    setLoading(true);
    // 2 start page transition
    nProgress.start();
    // 3 create a payment method via stripe (Token combacks here if successful)
    const { error, paymentMethod } = await stripe.createPaymentMethod({
      type: 'card',
      // it knows to render out only one card element, so we don't need to use ref
      card: elements.getElement(CardElement),
    });
    console.log(paymentMethod);
    // 4 handle any errors from stripe
    if (error) {
      setError(error);
      nProgress.done();
      return; // stops the checkout from happening
    }
    // 5 send token from step 3 to our keystone server via a custom mutation
    const order = await checkout({
      variables: {
        token: paymentMethod.id,
      },
    });
    console.log('Finished with the order');
    console.log('ORDER', order);
    // 6 change the pate to view the order
    // we are rerouting user to page /order and passing a query string with it
    router.push({
      // we specify that pathname will look for variable id which will be populated by query string we are providing. This is a Next.js way to passing variables
      pathname: '/order/[id]',
      query: { id: order.data.checkout.id },
    });
    // 7 close the cart
    closeCart();
    // 8 turn the loader off
    setLoading(false);
    nProgress.done();
  }
  return (
    <CheckoutFormStyles onSubmit={handleSubmit}>
      {error && <p style={{ fontSize: 16, color: 'red' }}>{error.message}</p>}
      {graphqlError && (
        <p style={{ fontSize: 16, color: 'red' }}>{graphqlError.message}</p>
      )}
      <CardElement />
      <SickButton>Check Out Now</SickButton>
    </CheckoutFormStyles>
  );
}

function CheckOut() {
  return (
    <Elements stripe={stripeLib}>
      <CheckOutForm />
    </Elements>
  );
}

export default CheckOut;
