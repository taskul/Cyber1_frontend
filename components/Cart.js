import styled from 'styled-components';
import CartStyles from './styles/CartStyles';
import Supreme from './styles/Supreme';
import CloseButton from './styles/CloseButton';
import { useUser } from './User';
import formatMoney from '../lib/formatMoney';
import calcTotalPrice from '../lib/calcTotalPrice';
import { useCart } from '../lib/cartState';
import RemoveFromCart from './RemoveFromCart';
import CheckOut from './CheckOut';

const CartItemStyles = styled.li`
  padding: 0.5rem 0;
  border-bottom: 1px solid var(--yellow, #e5d283);
  display: grid;
  grid-template-columns: auto 1fr auto;
  img {
    margin-right: 1rem;
  }
  h3,
  p {
    margin: 0px;
  }
`;

const FooterStyle = styled.div`
  background-color: var(--yellow);
  font-size: 3rem;
  font-weight: bold;
  color: #000;
`;

function CartItem({ cartItem }) {
  const { product } = cartItem;
  if (!product) return null;
  return (
    <CartItemStyles>
      <img
        width="100"
        src={product.photo.image.publicUrlTransformed}
        alt={product.name}
      />
      <div>
        <h3>{product.name}</h3>
        <p>
          {formatMoney(product.price * cartItem.quantity)} -{' '}
          <em>
            {cartItem.quantity} &times; {formatMoney(product.price)} each
          </em>
        </p>
      </div>
      <RemoveFromCart id={cartItem.id} />
    </CartItemStyles>
  );
}

export default function Cart() {
  const me = useUser();
  const { cartOpen, closeCart } = useCart();
  if (!me) return null;
  return (
    <CartStyles open={cartOpen}>
      <header>
        <Supreme>{me.name}'s Cart</Supreme>
      </header>
      <CloseButton onClick={closeCart}>&times;</CloseButton>
      <ul>
        {me.cart.map((cartItem) => (
          <CartItem key={cartItem.id} cartItem={cartItem} />
        ))}
      </ul>
      <FooterStyle>
        <p>Total: {formatMoney(calcTotalPrice(me.cart))}</p>
        <CheckOut />
      </FooterStyle>
    </CartStyles>
  );
}
