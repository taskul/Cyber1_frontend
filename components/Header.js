import Link from 'next/link';
import styled from 'styled-components';
import { useEffect, useState } from 'react';
import Nav from './Nav';
import Cart from './Cart';
import Search from './Search';

// tag template literal
// we are defining our CSS in this JavaScript component and we'll define our CSS withing componets to make it easier to find everything related to each component
// because Logo component wraps around <Link> component we can edit it's CSS by editing <a> tag shown below
// so we can easily change CSS of descendants of the component
const Logo = styled.h1`
  position: relative;
  margin-left: 2rem;
  background: var(--blackBg, #001524);
  border: 2px solid var(--yellow, yellow);
  font-size: 4rem;
  z-index: 2;
  transform: skew(-7deg);
  transition: background-color 1s, box-shadow 1s;
  :hover {
    background-color: var(--yellow, yellow);
    box-shadow: 0 0 5px #a0a14a, 0 0 10px #a0a14a, 0 0 25px #a0a14a,
      0 0 200px #a0a14a;
    -webkit-box-reflect: below 1px linear-gradient(transparent, #0005);
  }
  a {
    color: black;
    text-decoration: none;
    text-transform: uppercase;
    padding: 0.5rem 1rem;
    -webkit-text-stroke: 0.1px #5cd2e6;
  }
`;

const HeaderStyles = styled.header`
  .bar {
    /* we'll define a --yellow variable, also fall back yellow is used as a second argument in case --yellow is not working */
    border-bottom: 1px solid var(--yellow, yellow);
    display: grid;
    grid-template-columns: auto 1fr;
    justify-content: space-between;
    align-items: center;
  }

  .sub-bar {
    display: grid;
    grid-template-columns: 1fr auto;
    border-bottom: 1px solid var(--yellow, yellow);
  }
`;

function ClientOnly({ children, ...delegated }) {
  const [hasMounted, setHasMounted] = useState(false);

  useEffect(() => {
    setHasMounted(true);
  }, []);

  if (!hasMounted) {
    return null;
  }

  return <div {...delegated}>{children}</div>;
}

// This is our header that renders NavBar, logo and search bar
export default function Header() {
  return (
    <HeaderStyles>
      <div className="bar">
        {/* Instead of this being an <h1> it is a Logo component that we created above */}
        <Logo>
          <Link href="/">Cyber-1</Link>
        </Logo>
        <Nav />
      </div>
      <div className="sub-bar">
        <ClientOnly>
          <Search>Search</Search>
        </ClientOnly>
      </div>
      <Cart />
    </HeaderStyles>
  );
}
