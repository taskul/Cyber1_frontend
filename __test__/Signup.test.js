import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import userEvent from '@testing-library/user-event';
import Signup, { SIGNUP_MUTATION } from '../components/SignUp';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeUser } from '../lib/testUtils';

const me = fakeUser(); // this will create a random fake name and email
const password = '12345678';

const mocks = [
  // Mutation mock, first we specify what query we are going to make
  // our query has the same structure as the query we make with our SignUp component
  {
    request: {
      query: SIGNUP_MUTATION,
      variables: {
        name: me.name,
        email: me.email,
        password,
      },
    },
    // then we specify the data we are going to get back
    result: {
      data: {
        createUser: {
          __typename: 'User',
          id: 'abc123',
          email: me.email,
          name: me.name,
        },
      },
    },
  },

  // Current user mock, we don't need to mock it since we don't automatically log the user in, but if we were then we would use this query
  //   {
  //     // This query checks if there is a user logged in already, and the result is the data that matches our fakeUser data we provided
  //     request: { query: CURRENT_USER_QUERY },
  //     result: { data: { authenticatedItem: me } },
  //   },
];

describe('<SignUp />', () => {
  it('renders and matchces snapshot', () => {
    const { container } = render(
      <MockedProvider>
        <Signup />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });
  it.skip('calls the mutation properly', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <Signup />
      </MockedProvider>
    );
    // type into the boxes
    // simulating user using the website
    // type means typing into the field, and we'll use "placeholder" as a way to find the field because that is what user would use to spot the field. Field label also would work
    await userEvent.type(screen.getByPlaceholderText('Name'), me.name);
    await userEvent.type(screen.getByPlaceholderText('Email'), me.email);
    await userEvent.type(screen.getByPlaceholderText('Password'), password);
    // debug();
    await userEvent.click(screen.getByText('Sign Up'));
    await screen.findByText(`Signed up with ${me.email}`);
    // debug();
  });
});
