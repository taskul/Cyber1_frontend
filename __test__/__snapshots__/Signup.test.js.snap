// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp /> renders and matchces snapshot 1`] = `
<div>
  <form
    class="Form-sc-1xszr8q-0 jndoER"
    method="POST"
  >
    <h2>
      Sing Up For An Account
    </h2>
    <fieldset>
      <label
        for="name"
      >
        Name
        <input
          autocomplete="name"
          name="name"
          placeholder="Name"
          type="name"
          value=""
        />
      </label>
      <label
        for="email"
      >
        Email
        <input
          autocomplete="email"
          name="email"
          placeholder="Email"
          type="email"
          value=""
        />
      </label>
      <label
        for="password"
      >
        Password
        <input
          autocomplete="password"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Sign Up
      </button>
    </fieldset>
  </form>
</div>
`;
